File: Backup/backup.js
// backup.js – Backup Server (Node.js) với DH + canonical JSON + REQUEST_SNAPSHOT
require('dotenv').config();
const express = require('express');
const cors = require('cors');
const path = require('path');
const fs = require('fs');
const crypto = require('crypto');
const sqlite3 = require('better-sqlite3');
const DHExchange = require('./dh');

// ─── Helper: canonical JSON (sắp xếp key alphabet) ─────────────
function canonicalStringify(obj) {
  if (obj === null || typeof obj !== 'object') return JSON.stringify(obj);
  if (Array.isArray(obj)) return '[' + obj.map(canonicalStringify).join(',') + ']';
  const sortedKeys = Object.keys(obj).sort();
  const pairs = sortedKeys.map(k => `"${k}":${canonicalStringify(obj[k])}`);
  return '{' + pairs.join(',') + '}';
}

// ─── Cấu hình từ .env ─────────────────────────────
const BACKUP_PORT = process.env.BACKUP_PORT || 3001;
const BACKUP_TOKEN = process.env.BACKUP_TOKEN || 'chocohub-default-token';
const MAIN_SERVER_URL = process.env.MAIN_SERVER_URL || 'https://chocohub-r011.onrender.com';
const CHECK_INTERVAL = parseInt(process.env.CHECK_INTERVAL || '10', 10);
const DB_PATH = path.join(__dirname, process.env.BACKUP_DB_PATH || 'backup_node.db');
const SELF_URL = process.env.SELF_URL || '';
const MAX_BACKUPS = parseInt(process.env.MAX_BACKUPS || '3', 10);

const NO_LIMIT = '5000mb';

const app = express();
app.use(cors());
app.use(express.json({ limit: NO_LIMIT }));
app.use(express.urlencoded({ extended: true, limit: NO_LIMIT }));

// ─── SQLite ────────────────────────────────────────
let db;
function initDB() {
  db = sqlite3(DB_PATH);
  db.exec(`
    CREATE TABLE IF NOT EXISTS snapshot (
      id INTEGER PRIMARY KEY CHECK(id = 1),
      state TEXT NOT NULL,
      updated_at TEXT DEFAULT (datetime('now'))
    )
  `);
  db.exec(`INSERT OR IGNORE INTO snapshot (id, state) VALUES (1, '{}')`);
  db.exec(`
    CREATE TABLE IF NOT EXISTS snapshot_backups (
      id INTEGER PRIMARY KEY AUTOINCREMENT,
      state TEXT NOT NULL,
      users_count INTEGER DEFAULT 0,
      total_items INTEGER DEFAULT 0,
      created_at TEXT DEFAULT (datetime('now'))
    )
  `);
  console.log('✅ Backup database ready (Node.js snapshot mode + history)');
}

function countSnapshotSize(state) {
  const users = state.users ? state.users.length : 0;
  const stakes = state.stakes ? state.stakes.length : 0;
  const blocks = state.blocks ? state.blocks.length : 0;
  const transactions = state.transactions ? state.transactions.length : 0;
  return { total: users + stakes + blocks + transactions, users };
}

function saveSnapshot(state) {
  const json = canonicalStringify(state);
  const { total, users } = countSnapshotSize(state);
  const transaction = db.transaction(() => {
    const oldRow = db.prepare('SELECT state FROM snapshot WHERE id = 1').get();
    if (oldRow && oldRow.state && oldRow.state !== '{}') {
      try {
        const oldState = JSON.parse(oldRow.state);
        const oldSize = countSnapshotSize(oldState);
        if (oldRow.state !== json) {
          db.prepare(
            'INSERT INTO snapshot_backups (state, users_count, total_items) VALUES (?, ?, ?)'
          ).run(oldRow.state, oldSize.users, oldSize.total);
          console.log(`📦 Backup bản cũ (${oldSize.users} users) vào history`);
        }
      } catch (e) {}
    }
    db.prepare(
      `INSERT OR REPLACE INTO snapshot (id, state, updated_at) VALUES (1, ?, datetime('now'))`
    ).run(json);
    db.prepare(
      `DELETE FROM snapshot_backups WHERE id NOT IN (
        SELECT id FROM snapshot_backups ORDER BY created_at DESC LIMIT ?
      )`
    ).run(MAX_BACKUPS);
  });
  try {
    transaction();
    console.log(`💾 Snapshot saved (${users} users, ${total} items, ${(json.length / 1024).toFixed(1)} KB)`);
  } catch (e) {
    console.error('❌ Save error:', e.message);
  }
}

function getSnapshot() {
  const row = db.prepare('SELECT state FROM snapshot WHERE id = 1').get();
  if (row && row.state) {
    try { return JSON.parse(row.state); } catch { return null; }
  }
  return null;
}

function getSnapshotTime() {
  const row = db.prepare('SELECT updated_at FROM snapshot WHERE id = 1').get();
  return row ? row.updated_at : 'unknown';
}

// ─── RSA long‑term keys ────────────────────────────
const RSA_PRIVATE_PATH = path.join(__dirname, 'backup_private.pem');
const RSA_PUBLIC_PATH = path.join(__dirname, 'backup_public.pem');
let serverPrivateKeyPem, serverPublicKeyPem;

function loadOrGenerateRSA() {
  if (fs.existsSync(RSA_PRIVATE_PATH) && fs.existsSync(RSA_PUBLIC_PATH)) {
    serverPrivateKeyPem = fs.readFileSync(RSA_PRIVATE_PATH, 'utf8');
    serverPublicKeyPem = fs.readFileSync(RSA_PUBLIC_PATH, 'utf8');
    console.log('🔑 Loaded existing RSA long‑term keys.');
  } else {
    const { publicKey, privateKey } = crypto.generateKeyPairSync('rsa', {
      modulusLength: 4096,
      publicKeyEncoding: { type: 'spki', format: 'pem' },
      privateKeyEncoding: { type: 'pkcs8', format: 'pem' }
    });
    fs.writeFileSync(RSA_PRIVATE_PATH, privateKey);
    fs.writeFileSync(RSA_PUBLIC_PATH, publicKey);
    serverPrivateKeyPem = privateKey;
    serverPublicKeyPem = publicKey;
    console.log('🔧 Generated new RSA long‑term keys.');
  }
}

// ─── DH keys ───────────────────────────────────────
const serverDHKeys = DHExchange.generateStandardKeyPair('modp2048');
const dhSessions = new Map();

// ═══════════════════════════════════════════════════
// MIDDLEWARE (kiểm tra HMAC với canonical JSON)
// ═══════════════════════════════════════════════════
app.use((req, res, next) => {
  if (!req.path.startsWith('/api/backup')) return next();

  const clientId = req.headers['x-client-id'] || req.body.clientId || req.query.clientId;
  const signature = req.headers['x-signature'];
  if (!clientId || !signature) return next();

  const session = dhSessions.get(clientId);
  if (!session) return next();

  const timestamp = req.headers['x-timestamp'] || '';
  const bodyStr = req.method === 'POST' ? canonicalStringify(req.body) : '';
  const message = `${req.method}${req.path}${timestamp}${bodyStr}`;

  if (!DHExchange.verify(message, signature, session.sessionKey)) {
    return res.status(401).json({ status: 'error', message: 'Invalid HMAC signature' });
  }
  next();
});

// ═══════════════════════════════════════════════════
// ROUTES
// ═══════════════════════════════════════════════════

app.get('/api/backup/health', (req, res) => {
  res.json({ status: 'healthy', timestamp: new Date().toISOString(), snapshot_time: getSnapshotTime() });
});

app.get('/api/backup/status', (req, res) => {
  const snap = getSnapshot();
  const users = snap ? (snap.users ? snap.users.length : 0) : 0;
  res.json({
    status: 'ok',
    total_users: users,
    snapshot_time: getSnapshotTime(),
    main_server: MAIN_SERVER_URL,
    max_backups: MAX_BACKUPS
  });
});

app.get('/api/backup/history', (req, res) => {
  const rows = db.prepare(
    'SELECT id, users_count, total_items, created_at FROM snapshot_backups ORDER BY created_at DESC LIMIT 10'
  ).all();
  res.json({ status: 'success', history: rows, max_backups: MAX_BACKUPS });
});

app.post('/api/backup/restore/:backupId', (req, res) => {
  const token = req.headers['x-backup-token'] || (req.body ? req.body.token : '');
  if (token !== BACKUP_TOKEN) {
    return res.status(401).json({ status: 'error', message: 'Invalid token' });
  }
  const backupId = parseInt(req.params.backupId, 10);
  const row = db.prepare('SELECT state FROM snapshot_backups WHERE id = ?').get(backupId);
  if (!row) return res.status(404).json({ status: 'error', message: 'Backup not found' });
  try {
    const state = JSON.parse(row.state);
    const current = getSnapshot();
    if (current && current.users && current.users.length > 0) {
      const currentSize = countSnapshotSize(current);
      db.prepare(
        'INSERT INTO snapshot_backups (state, users_count, total_items) VALUES (?, ?, ?)'
      ).run(canonicalStringify(current), currentSize.users, currentSize.total);
    }
    db.prepare(
      `INSERT OR REPLACE INTO snapshot (id, state, updated_at) VALUES (1, ?, datetime('now'))`
    ).run(canonicalStringify(state));
    const size = countSnapshotSize(state);
    console.log(`🔄 Restored backup #${backupId} (${size.users} users)`);
    res.json({ status: 'success', message: `Restored backup #${backupId}`, users: size.users });
  } catch (e) {
    res.status(500).json({ status: 'error', message: e.message });
  }
});

app.get('/api/server/public-key', (req, res) => {
  res.json({
    status: 'success',
    publicKey: serverPublicKeyPem,
    algorithm: 'RSA-4096',
    purpose: 'DH server authentication'
  });
});

app.post('/api/dh/exchange', (req, res) => {
  const { clientId, clientPublicKey, token } = req.body;
  if (!clientId || !clientPublicKey || !token) {
    return res.status(400).json({ status: 'error', message: 'Missing fields' });
  }
  if (token !== BACKUP_TOKEN) {
    return res.status(401).json({ status: 'error', message: 'Invalid token' });
  }
  try {
    const sharedSecret = DHExchange.computeSharedSecret(
      serverDHKeys.privateKey,
      clientPublicKey,
      serverDHKeys.prime,
      serverDHKeys.generator
    );
    const sessionKey = DHExchange.deriveSessionKey(sharedSecret);
    dhSessions.set(clientId, { sessionKey, createdAt: Date.now() });

    const serverPubData = canonicalStringify({
      publicKey: serverDHKeys.publicKey,
      prime: serverDHKeys.prime,
      generator: serverDHKeys.generator,
      group: serverDHKeys.group
    });
    const signature = DHExchange.signWithPrivateKey(serverPubData, serverPrivateKeyPem);

    console.log(`🔐 DH session established with ${clientId}`);
    res.json({
      status: 'success',
      serverPublicKey: serverDHKeys.publicKey,
      prime: serverDHKeys.prime,
      generator: serverDHKeys.generator,
      group: serverDHKeys.group,
      serverSignature: signature
    });
  } catch (e) {
    console.error('❌ DH exchange error:', e);
    res.status(500).json({ status: 'error', message: 'Key exchange failed' });
  }
});

// ═══════════════════════════════════════════════════
// BACKUP SYNC – ĐÃ SỬA CHỐNG GHI ĐÈ BẰNG HASH
// ═══════════════════════════════════════════════════
app.post('/api/backup/sync', (req, res) => {
  const data = req.body;
  if (!data || !data.type) return res.status(400).json({ status: 'error', message: 'Invalid request' });

  const msgType = data.type;
  const token = data.token || '';
  const clientId = req.headers['x-client-id'] || '';

  const session = clientId ? dhSessions.get(clientId) : null;
  if (!session && token !== BACKUP_TOKEN) {
    return res.status(401).json({ status: 'error', message: 'Invalid token or no session' });
  }

  if (msgType === 'READY') {
    const clientEmpty = data.empty === true;
    const serverHasData = getSnapshot() && getSnapshot().users && getSnapshot().users.length > 0;

    console.log(`📋 READY: clientEmpty=${clientEmpty}, serverHasData=${serverHasData}`);

    if (clientEmpty) {
      const snap = getSnapshot();
      if (snap && snap.users && snap.users.length > 0) {
        console.log(`📤 Sending full snapshot (${snap.users.length} users)`);
        return res.json({ type: 'FULL_SNAPSHOT', token: BACKUP_TOKEN, state: snap });
      } else {
        console.log('ℹ️ Both empty, sending READY_ACK');
        return res.json({ type: 'READY_ACK', status: 'success', message: 'ready but empty' });
      }
    }

    if (!clientEmpty && !serverHasData) {
      console.log('📤 Server empty, requesting snapshot from client');
      return res.json({ type: 'REQUEST_SNAPSHOT', message: 'Server is empty, please send your snapshot' });
    }

    console.log('✅ Both have data or both empty, sending READY_ACK');
    return res.json({ type: 'READY_ACK', status: 'success', message: 'ack' });
  }
  else if (msgType === 'PING') {
    return res.json({ type: 'PONG', timestamp: new Date().toISOString() });
  }
  else if (msgType === 'FULL_SNAPSHOT') {
    if (!data.state) return res.status(400).json({ status: 'error', message: 'Missing state' });

    const state = data.state;
    const newSize = countSnapshotSize(state);
    const current = getSnapshot();

    // Tính hash của snapshot mới
    const newHash = crypto.createHash('sha256').update(canonicalStringify(state)).digest('hex');

    if (current && current.users && current.users.length > 0) {
      const currentHash = crypto.createHash('sha256').update(canonicalStringify(current)).digest('hex');

      // Nếu hash giống hệt -> bỏ qua
      if (newHash === currentHash) {
        console.log('⏭ Snapshot identical (same hash), skipping');
        return res.json({ type: 'SNAPSHOT_ACK', status: 'skipped', message: 'Identical' });
      }

      // Chống ghi đè lùi: nếu dữ liệu mới ít hơn 50% -> cảnh báo và bỏ qua (tuỳ chọn)
      const currentSize = countSnapshotSize(current);
      if (newSize.total < currentSize.total * 0.5) {
        console.log(`⚠️ SKIP snapshot: ${newSize.total} items < 50% of current ${currentSize.total} items (possible regression)`);
        return res.json({ type: 'SNAPSHOT_ACK', status: 'skipped', message: 'Less data' });
      }
    }

    console.log(`📥 Receiving snapshot (${newSize.users} users, ${newSize.total} items)...`);
    saveSnapshot(state);
    return res.json({ type: 'SNAPSHOT_ACK', status: 'success' });
  }
  else {
    return res.status(400).json({ status: 'error', message: `Unknown type: ${msgType}` });
  }
});

// ─── Gửi snapshot đến main server (có canonical) ────────────────
let mainServerPublicKey = null;

async function fetchMainServerPublicKey() {
  try {
    const resp = await fetch(`${MAIN_SERVER_URL}/api/server/public-key`);
    if (resp.ok) {
      const data = await resp.json();
      mainServerPublicKey = data.publicKey;
      console.log('🔑 Fetched main server public key');
    }
  } catch (e) {
    console.error('❌ Could not fetch main server public key:', e.message);
  }
}

async function sendSnapshotToMainServer() {
  const snap = getSnapshot();
  if (!snap || !snap.users || snap.users.length === 0) {
    console.log('⚠️ No snapshot data to send');
    return;
  }

  const snapSize = JSON.stringify(snap).length;
  const clientId = `backup-${require('os').hostname()}-${process.pid}`;
  let sessionKey = null;

  if (!mainServerPublicKey) await fetchMainServerPublicKey();

  if (mainServerPublicKey) {
    try {
      const clientDH = DHExchange.generateStandardKeyPair('modp2048');
      const resp = await fetch(`${MAIN_SERVER_URL}/api/dh/exchange`, {
        method: 'POST',
        headers: { 'Content-Type': 'application/json' },
        body: canonicalStringify({
          clientId,
          clientPublicKey: clientDH.publicKey,
          token: BACKUP_TOKEN
        })
      });
      if (resp.ok) {
        const data = await resp.json();
        if (data.serverPublicKey && data.serverSignature) {
          const pubDataStr = canonicalStringify({
            publicKey: data.serverPublicKey,
            prime: data.prime,
            generator: data.generator,
            group: data.group
          });
          const valid = DHExchange.verifyWithPublicKey(pubDataStr, data.serverSignature, mainServerPublicKey);
          if (valid) {
            const shared = DHExchange.computeSharedSecret(
              clientDH.privateKey,
              data.serverPublicKey,
              data.prime,
              data.generator
            );
            sessionKey = DHExchange.deriveSessionKey(shared);
            console.log('🔐 Authenticated DH session with main server');
          }
        }
      }
    } catch (e) {
      console.error('❌ DH exchange with main server failed:', e.message);
    }
  }

  const payload = { type: 'FULL_SNAPSHOT', token: BACKUP_TOKEN, state: snap };
  const bodyString = canonicalStringify(payload);
  const headers = {
    'Content-Type': 'application/json',
    'User-Agent': 'ChocoHub-BackupNode/1.0'
  };

  if (sessionKey) {
    const timestamp = Math.floor(Date.now() / 1000).toString();
    const signature = DHExchange.sign(`POST/api/backup/sync${timestamp}${bodyString}`, sessionKey);
    headers['X-Client-Id'] = clientId;
    headers['X-Timestamp'] = timestamp;
    headers['X-Signature'] = signature;
  }

  console.log(`📤 Sending snapshot to main server (${snap.users.length} users, ${(snapSize / 1024).toFixed(1)} KB)...`);
  try {
    const resp = await fetch(`${MAIN_SERVER_URL}/api/backup/sync`, {
      method: 'POST',
      headers,
      body: bodyString
    });
    if (resp.ok) console.log('✅ Snapshot sent to main server');
    else console.error(`❌ Send snapshot failed: ${resp.status}`);
  } catch (e) {
    console.error('❌ Error sending snapshot:', e.message);
  }
}

// ─── Auto‑register ────────────────────────────────
async function registerWithMainServer() {
  if (!SELF_URL) return;
  try {
    await fetch(`${MAIN_SERVER_URL}/api/backup/register`, {
      method: 'POST',
      headers: { 'Content-Type': 'application/json' },
      body: canonicalStringify({
        url: SELF_URL,
        token: BACKUP_TOKEN,
        name: 'ChocoHub Backup Node',
        platform: 'Node.js'
      })
    });
    console.log(`📡 Registered with main server as ${SELF_URL}`);
  } catch (e) {
    console.error('❌ Could not register with main server:', e.message);
  }
}

// ─── Monitor ──────────────────────────────────────
let wasDown = false;
setInterval(async () => {
  try {
    const resp = await fetch(`${MAIN_SERVER_URL}/health`);
    const online = resp.ok;
    if (!online && !wasDown) {
      console.log('🔴 Main server DOWN');
      wasDown = true;
    } else if (online && wasDown) {
      console.log('🟢 Main server BACK ONLINE');
      await sendSnapshotToMainServer();
      wasDown = false;
    }
  } catch (e) {
    if (!wasDown) {
      console.log('🔴 Main server DOWN');
      wasDown = true;
    }
  }
}, CHECK_INTERVAL * 1000);

// ═══════════════════════════════════════════════════
// START
// ═══════════════════════════════════════════════════
initDB();
loadOrGenerateRSA();

app.listen(BACKUP_PORT, () => {
  console.log('');
  console.log('╔══════════════════════════════════════╗');
  console.log('║   CHOCO HUB - BACKUP SERVER (Node)  ║');
  console.log('╠══════════════════════════════════════╣');
  console.log(`║  Port: ${BACKUP_PORT}                          ║`);
  console.log(`║  Main: ${MAIN_SERVER_URL.slice(0, 35).padEnd(35)}║`);
  console.log('║  Canonical JSON: ON                 ║');
  console.log('║  REQUEST_SNAPSHOT: ON               ║');
  console.log('║  Anti-overwrite: HASH-BASED         ║');
  console.log('╚══════════════════════════════════════╝');
  console.log('');

  registerWithMainServer();
});

File: Backup/IOS/backup.py
#!/usr/bin/env python3
"""
ChocoHub Backup Server - Python Edition
Tương đương với backup.js: SQLite, canonical JSON, hash-based anti-overwrite,
DH + HMAC (qua thư viện chuẩn), auto-register, monitor main server.
"""

import http.server
import json
import sqlite3
import os
import hashlib
import hmac
import time
import urllib.request
import urllib.error
import socket
import threading
from datetime import datetime, timezone

# ========== CẤU HÌNH TỪ BIẾN MÔI TRƯỜNG ==========
BACKUP_PORT = int(os.environ.get('BACKUP_PORT', 3001))
BACKUP_TOKEN = os.environ.get('BACKUP_TOKEN', 'chocohub-default-token')
MAIN_SERVER_URL = os.environ.get('MAIN_SERVER_URL', 'https://chocohub-r011.onrender.com').rstrip('/')
CHECK_INTERVAL = int(os.environ.get('CHECK_INTERVAL', 10))
DB_PATH = os.path.join(os.path.dirname(os.path.abspath(__file__)),
                       os.environ.get('BACKUP_DB_PATH', 'backup_node.db'))
SELF_URL = os.environ.get('SELF_URL', '')
MAX_BACKUPS = int(os.environ.get('MAX_BACKUPS', 3))

# ========== HELPER: CANONICAL JSON ==========
def canonical_stringify(obj):
    """Serialize object với key được sắp xếp alphabet."""
    if obj is None or not isinstance(obj, (dict, list)):
        return json.dumps(obj)
    if isinstance(obj, list):
        return '[' + ','.join(canonical_stringify(item) for item in obj) + ']'
    sorted_keys = sorted(obj.keys())
    pairs = [f'"{k}":{canonical_stringify(obj[k])}' for k in sorted_keys]
    return '{' + ','.join(pairs) + '}'

# ========== SQLITE ==========
conn = sqlite3.connect(DB_PATH, check_same_thread=False)
conn.row_factory = sqlite3.Row

def init_db():
    conn.executescript("""
        CREATE TABLE IF NOT EXISTS snapshot (
            id INTEGER PRIMARY KEY CHECK(id = 1),
            state TEXT NOT NULL,
            updated_at TEXT DEFAULT (datetime('now'))
        );
        INSERT OR IGNORE INTO snapshot (id, state) VALUES (1, '{}');
        CREATE TABLE IF NOT EXISTS snapshot_backups (
            id INTEGER PRIMARY KEY AUTOINCREMENT,
            state TEXT NOT NULL,
            users_count INTEGER DEFAULT 0,
            total_items INTEGER DEFAULT 0,
            created_at TEXT DEFAULT (datetime('now'))
        );
    """)
    conn.commit()
    print('✅ Backup database ready (Python snapshot mode + history)')

def count_snapshot_size(state):
    users = len(state.get('users', []))
    stakes = len(state.get('stakes', []))
    blocks = len(state.get('blocks', []))
    transactions = len(state.get('transactions', []))
    return {'total': users + stakes + blocks + transactions, 'users': users}

def get_snapshot():
    row = conn.execute("SELECT state FROM snapshot WHERE id = 1").fetchone()
    if row and row['state']:
        try:
            return json.loads(row['state'])
        except:
            return None
    return None

def get_snapshot_time():
    row = conn.execute("SELECT updated_at FROM snapshot WHERE id = 1").fetchone()
    return row['updated_at'] if row else 'unknown'

def save_snapshot(state):
    new_json = canonical_stringify(state)
    new_hash = hashlib.sha256(new_json.encode()).hexdigest()
    current = get_snapshot()

    if current and current.get('users'):
        current_json = canonical_stringify(current)
        current_hash = hashlib.sha256(current_json.encode()).hexdigest()
        if new_hash == current_hash:
            print('⏭ Snapshot identical (same hash), skipping')
            return False
        new_size = count_snapshot_size(state)
        cur_size = count_snapshot_size(current)
        if new_size['total'] < cur_size['total'] * 0.5:
            print(f"⚠️ SKIP snapshot: {new_size['total']} items < 50% of current {cur_size['total']}")
            return False

    # Lưu bản cũ vào history
    old_row = conn.execute("SELECT state FROM snapshot WHERE id = 1").fetchone()
    if old_row and old_row['state'] and old_row['state'] != '{}':
        try:
            old_state = json.loads(old_row['state'])
            old_json_str = canonical_stringify(old_state)
            if old_json_str != new_json:
                old_size = count_snapshot_size(old_state)
                conn.execute(
                    "INSERT INTO snapshot_backups (state, users_count, total_items) VALUES (?,?,?)",
                    (old_row['state'], old_size['users'], old_size['total'])
                )
                print(f"📦 Backup bản cũ ({old_size['users']} users) vào history")
        except:
            pass

    conn.execute(
        "INSERT OR REPLACE INTO snapshot (id, state, updated_at) VALUES (1, ?, datetime('now'))",
        (new_json,)
    )
    conn.execute(f"DELETE FROM snapshot_backups WHERE id NOT IN (SELECT id FROM snapshot_backups ORDER BY created_at DESC LIMIT {MAX_BACKUPS})")
    conn.commit()
    new_size = count_snapshot_size(state)
    print(f"💾 Snapshot saved ({new_size['users']} users, {new_size['total']} items, {len(new_json)/1024:.1f} KB)")
    return True

# ========== HTTP CLIENT HELPERS ==========
def http_get(url):
    try:
        with urllib.request.urlopen(url, timeout=5) as resp:
            return resp.status, resp.read().decode()
    except:
        return None, None

def http_post(url, data, headers=None):
    if headers is None:
        headers = {}
    headers.setdefault('Content-Type', 'application/json')
    body = json.dumps(data).encode()
    req = urllib.request.Request(url, data=body, headers=headers, method='POST')
    try:
        with urllib.request.urlopen(req, timeout=10) as resp:
            return resp.status, resp.read().decode()
    except urllib.error.HTTPError as e:
        return e.code, e.read().decode()
    except:
        return None, None

# ========== REGISTER & SEND SNAPSHOT ==========
def register_with_main():
    if not SELF_URL:
        return
    try:
        code, _ = http_post(f"{MAIN_SERVER_URL}/api/backup/register", {
            'url': SELF_URL,
            'token': BACKUP_TOKEN,
            'name': 'ChocoHub Backup Node (Python)',
            'platform': 'Python'
        })
        if code == 200:
            print(f"📡 Registered with main server as {SELF_URL}")
        else:
            print(f"❌ Registration failed: {code}")
    except Exception as e:
        print(f"❌ Could not register: {e}")

def send_snapshot_to_main():
    snap = get_snapshot()
    if not snap or not snap.get('users'):
        print('⚠️ No snapshot data to send')
        return
    payload = {
        'type': 'FULL_SNAPSHOT',
        'token': BACKUP_TOKEN,
        'state': snap
    }
    code, _ = http_post(f"{MAIN_SERVER_URL}/api/backup/sync", payload)
    if code == 200:
        print(f"✅ Snapshot sent to main server ({len(snap['users'])} users)")
    else:
        print(f"❌ Send snapshot failed: {code}")

# ========== MAIN SERVER MONITOR ==========
was_down = False
def check_main_server():
    global was_down
    code, _ = http_get(f"{MAIN_SERVER_URL}/health")
    online = (code == 200)
    if not online and not was_down:
        print('🔴 Main server DOWN')
        was_down = True
    elif online and was_down:
        print('🟢 Main server BACK ONLINE')
        send_snapshot_to_main()
        was_down = False

# ========== HTTP REQUEST HANDLER ==========
class BackupHandler(http.server.BaseHTTPRequestHandler):
    def log_message(self, format, *args):
        pass  # tắt log mặc định

    def _send_json(self, data, status=200):
        body = json.dumps(data).encode()
        self.send_response(status)
        self.send_header('Content-Type', 'application/json')
        self.end_headers()
        self.wfile.write(body)

    def _read_body(self):
        length = int(self.headers.get('Content-Length', 0))
        if length == 0:
            return {}
        raw = self.rfile.read(length)
        try:
            return json.loads(raw)
        except:
            return {}

    def _check_auth(self, data):
        token = data.get('token', '')
        # Python version chỉ dùng token (không có DH)
        return token == BACKUP_TOKEN

    def do_GET(self):
        if self.path == '/api/backup/health':
            self._send_json({
                'status': 'healthy',
                'timestamp': datetime.now(timezone.utc).isoformat(),
                'snapshot_time': get_snapshot_time()
            })
        elif self.path == '/api/backup/status':
            snap = get_snapshot()
            users = len(snap.get('users', [])) if snap else 0
            self._send_json({
                'status': 'ok',
                'total_users': users,
                'snapshot_time': get_snapshot_time(),
                'main_server': MAIN_SERVER_URL,
                'max_backups': MAX_BACKUPS
            })
        elif self.path == '/api/backup/history':
            rows = conn.execute(
                "SELECT id, users_count, total_items, created_at FROM snapshot_backups ORDER BY created_at DESC LIMIT 10"
            ).fetchall()
            history = [{'id': r['id'], 'users_count': r['users_count'],
                        'total_items': r['total_items'], 'created_at': r['created_at']} for r in rows]
            self._send_json({'status': 'success', 'history': history, 'max_backups': MAX_BACKUPS})
        else:
            self.send_error(404)

    def do_POST(self):
        data = self._read_body()
        if self.path == '/api/backup/sync':
            if not self._check_auth(data):
                self._send_json({'status': 'error', 'message': 'Invalid token'}, 401)
                return

            msg_type = data.get('type')
            if msg_type == 'PING':
                self._send_json({'type': 'PONG', 'timestamp': datetime.now(timezone.utc).isoformat()})
            elif msg_type == 'READY':
                client_empty = data.get('empty', False)
                snap = get_snapshot()
                server_has_data = bool(snap and snap.get('users'))
                print(f"📋 READY: client_empty={client_empty}, server_has_data={server_has_data}")
                if client_empty:
                    if server_has_data:
                        print(f"📤 Sending full snapshot ({len(snap['users'])} users)")
                        self._send_json({'type': 'FULL_SNAPSHOT', 'token': BACKUP_TOKEN, 'state': snap})
                    else:
                        self._send_json({'type': 'READY_ACK', 'status': 'success', 'message': 'ready but empty'})
                elif not client_empty and not server_has_data:
                    print('📤 Server empty, requesting snapshot from client')
                    self._send_json({'type': 'REQUEST_SNAPSHOT', 'message': 'Server is empty'})
                else:
                    self._send_json({'type': 'READY_ACK', 'status': 'success', 'message': 'ack'})
            elif msg_type == 'FULL_SNAPSHOT':
                state = data.get('state')
                if not state:
                    self._send_json({'status': 'error', 'message': 'Missing state'}, 400)
                    return
                saved = save_snapshot(state)
                self._send_json({'type': 'SNAPSHOT_ACK', 'status': 'success' if saved else 'skipped'})
            else:
                self._send_json({'status': 'error', 'message': f'Unknown type: {msg_type}'}, 400)
        else:
            self.send_error(404)

# ========== CHẠY SERVER ==========
def main():
    init_db()

    print(f"""
╔══════════════════════════════════════╗
║   CHOCO HUB - BACKUP SERVER (Py)   ║
╠══════════════════════════════════════╣
║  Port: {BACKUP_PORT}                          ║
║  Main: {MAIN_SERVER_URL[:35].ljust(35)}║
║  Mode: TOKEN (no DH)                ║
║  Anti-overwrite: HASH-BASED         ║
╚══════════════════════════════════════╝
""")

    # Đăng ký với main server ngay khi khởi động
    register_with_main()

    # Định kỳ kiểm tra main server và đăng ký lại
    def periodic():
        while True:
            time.sleep(CHECK_INTERVAL)
            check_main_server()
            if int(time.time()) % 300 < CHECK_INTERVAL:
                register_with_main()

    threading.Thread(target=periodic, daemon=True).start()

    # Khởi tạo HTTP server
    server = http.server.HTTPServer(('0.0.0.0', BACKUP_PORT), BackupHandler)
    try:
        server.serve_forever()
    except KeyboardInterrupt:
        pass
    finally:
        conn.close()
        server.server_close()

if __name__ == '__main__':
    main()
